java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -162,6 +164,23 @@ abstract class AbstractArcHost(
         entityHandleManager = entityHandleManager(arcId)
     )
 
+    override suspend fun setOnArcStateChange(arcId: ArcId, block: (ArcId, ArcState) -> Unit) {
+        lookupOrCreateArcHostContext(arcId.toString()).onArcStateChangeCallback.add(block)
+    }
+
+    private fun setArcState(context: ArcHostContext, state: ArcState) {
+        context.arcState = state
+        context.onArcStateChangeCallback.forEach {
+            try {
+                it(context.arcId.toArcId(), state)
+            } catch (e: Exception) {
+                log.debug(e) {
+                    "Exception in onArcStateChangeCallback for ${context.arcId}"
+                }
+            }
+        }
```

Comment:
I feel like this should probably be managed by the ArcHostContext itself, not the AbstractArcHost... it's leaking implementation details of the context to AbstractArcHost unnecessarily.